pkg/log/log_file.go

```diff
@@ -0,0 +1,194 @@
+
+// Package main under directory cmd parses and validates user input,
+// instantiates and initializes objects imported from pkg, and runs
+// the process.
+package log
+
+import (
+	"os"
+	"path/filepath"
+	"sort"
+	"strings"
+	"time"
+
+	"github.com/spf13/pflag"
+
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/klog"
+)
+
+const (
+	logToStdErrFlag = "logtostderr"
+	logDirFlag      = "log_dir"
+	logFileFlag     = "log_file"
+	maxSizeFlag     = "log_file_max_size"
+	maxNumFlag      = "log_file_max_num"
+
+	// Check log file number every 10 mins.
+	logFileCheckInterval = time.Minute * 10
+)
+
+var (
+	maxNumArg     = uint16(0)
+	logFileMaxNum = uint16(0)
+	logDir        = ""
+
+	executableName = filepath.Base(os.Args[0])
+)
+
+func AddFlags(fs *pflag.FlagSet) {
+	fs.Uint16Var(&maxNumArg, maxNumFlag, maxNumArg, "Maximum number of log files per severity level to be kept. Value 0 means unlimited.")
+}
+
+// InitLogFileLimits initializes log file maximum size and maximum number limits based on the
+// command line flags.
+func InitLogFileLimits(fs *pflag.FlagSet) {
+	var err error
+	var logToStdErr bool
+	var logFile string
+	var maxSize uint64
+
+	logToStdErr, err = fs.GetBool(logToStdErrFlag)
+	if err != nil {
+		// Should not happen. Return for safety.
+		return
+	}
+	if logToStdErr {
+		// Logging to files is not enabled.
+		return
+	}
+
+	logFile, err = fs.GetString(logFileFlag)
+	if err != nil {
+		return
+	}
+	if logFile != "" {
+		// Log to a single file. klog will take care of the max size limit.
+		return
+	}
+
+	maxSize, err = fs.GetUint64(maxSizeFlag)
+	if err != nil {
+		return
+	}
+	if maxSize > 1024*1024 {
+		klog.Errorf("The specified log file max size %d is too big, ignored", maxSize)
+	} else {
+		maxSize = maxSize * 1024 * 1024
+
+		// klog does not respect the max file size specified by --log_file_max_size
+		// when --log_file is not used. Here as a workaround, we directly set the
+		// specified max size to klog.MaxSize.
+		if klog.MaxSize != maxSize {
+			klog.MaxSize = maxSize
+			klog.Infof("Set log file max size to %d", maxSize)
+		}
+	}
+
+	if maxNumArg > 0 {
+		logDir, err = fs.GetString(logDirFlag)
+		if err != nil {
+			return
+		}
+
+		logFileMaxNum = maxNumArg
+		if logDir == "" {
+			// Log to the tmp dir.
+			logDir = os.TempDir()
+		}
+	}
+}
+
+// StartLogFileNumberMonitor starts monitoring the log files to make sure the
+// number of log files does not exceed the maximum limit, when the log file
+// number limit is configured.
+func StartLogFileNumberMonitor(stopCh <-chan struct{}) {
+	if logFileMaxNum == 0 {
+		// The maximum log file number limit is not configured.
+		return
+	}
+
+	go func() {
+		klog.Infof("Starting log file monitoring. Maximum log file number is %d", logFileMaxNum)
+		wait.Until(checkLogFiles, logFileCheckInterval, stopCh)
+	}()
+}
+
+func checkLogFiles() {
+	f, err := os.Open(logDir)
+	if err != nil {
+		klog.Errorf("Failed to open log directory %s: %v", logDir, err)
+		return
+	}
```

Comment:
`defer f.Close()` here in case of Readdir failure?
The error of deferred call is normally ignored.
https://github.com/kubernetes/kubernetes/blob/a138be8722ebd0ce281029fa747315500a99ffd5/pkg/kubelet/kuberuntime/logs/logs.go#L295
https://github.com/moby/moby/blob/3b4cfa97237a8e1fb5eb985e4a7c0717cd14f5c8/daemon/graphdriver/devmapper/deviceset.go#L556

Reply:
Yes. I hope to close the file earlier. Changed the code to cover Readdir failure.